bench/lib/tile_parser.js

```diff
@@ -0,0 +1,141 @@
+// @flow
+
+import Protobuf from 'pbf';
+import VT from '@mapbox/vector-tile';
+import assert from 'assert';
+
+import deref from '../../src/style-spec/deref';
+import Style from '../../src/style/style';
+import { Evented } from '../../src/util/evented';
+import { normalizeSourceURL, normalizeTileURL } from '../../src/util/mapbox';
+import WorkerTile from '../../src/source/worker_tile';
+import StyleLayerIndex from '../../src/style/style_layer_index';
+
+import type { StyleSpecification } from '../../src/style-spec/types';
+import type { WorkerTileResult } from '../../src/source/worker_source';
+import type { OverscaledTileID } from '../../src/source/tile_id';
+import type { TileJSON } from '../../src/types/tilejson';
+
+class StubMap extends Evented {
+    _transformRequest(url) {
+        return {url};
+    }
+}
+
+function createStyle(styleJSON: StyleSpecification): Promise<Style> {
+    return new Promise((resolve, reject) => {
+        const style = new Style((new StubMap(): any));
+        style.loadJSON(styleJSON);
+        style
+            .on('style.load', () => resolve(style))
+            .on('error', reject);
+    });
+}
+
+function fetchTileJSON(sourceURL: string): Promise<TileJSON> {
+    return fetch(normalizeSourceURL(sourceURL))
+        .then(response => response.json());
+}
+
+export default class TileParser {
+    styleJSON: StyleSpecification;
+    tileJSON: TileJSON;
+    sourceID: string;
+    layerIndex: StyleLayerIndex;
+    icons: Object;
+    glyphs: Object;
+    style: Style;
+    actor: { send: Function };
+
+    constructor(styleJSON: StyleSpecification, sourceID: string) {
+        this.styleJSON = styleJSON;
+        this.sourceID = sourceID;
+        this.layerIndex = new StyleLayerIndex(deref(this.styleJSON.layers));
+        this.glyphs = {};
+        this.icons = {};
+    }
+
+    loadImages(params: Object, callback: Function) {
+        const key = JSON.stringify(params);
+        if (this.icons[key]) {
+            callback(null, this.icons[key]);
+        } else {
+            this.style.getImages('', params, (err, icons) => {
```

Comment:
Good point — although it didn't show any noticeable effect (since it only affects one iteration), I added tile parsing in setup to prepopulate the cache just in case.